From 497a76829615a587045f8e941a8693ff24ecc20f Mon Sep 17 00:00:00 2001
From: Steve Best <sbest@redhat.com>
Date: Thu, 21 Aug 2014 18:45:03 -0400
Subject: [idle] intel_idle: Broadwell support

Message-id: <20140821184502.26640.71669.sendpatchset@ibm-p7r2-01.lab.bos.redhat.com>
Patchwork-id: 88687
O-Subject: [PATCH RHEL7.1 BZ1086453] intel_idle: Broadwell support
Bugzilla: 1086453
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Mark Langsdorf <mlangsdo@redhat.com>

RHBZ#:
------
https://bugzilla.redhat.com/show_bug.cgi?id=1086453

Description:
------------
Broadwell (BDW) is similar to Haswell (HSW), the preceding processor generation.

Currently, the only difference in their C-state tables is that PC3 max exit latency
is 33usec on HSW and 40usec on BDW.

Signed-off-by: Len Brown <len.brown@intel.com>

RHEL Version Found:
-------------------
RHEL 7.0

Brew:
-----
https://brewweb.devel.redhat.com/taskinfo?taskID=7854795

Jarod: This was built on top of bz 1048050 which has already been posted
http://post-office.corp.redhat.com/archives/rhkernel-list/2014-August/msg00854.html

Upstream:
---------
http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=a138b56800f4b83a7af69a9958d04f0f124eb37b

Test Status:
------------
Tested successfully by me using intel-sharkbay-ult-02 which is CPU family 6 model 61.

before the patch
dmesg | grep intel_idle
[    1.961073] intel_idle: does not run on family 6 model 61

after the patch
[    1.954825] intel_idle: MWAIT substates: 0x11142120
[    1.954827] intel_idle: v0.4 model 0x3D
[    1.954829] intel_idle: lapic_timer_reliable_states 0xffffffff

cat /sys/devices/system/cpu/cpuidle/current_drive
intel_idle

cat /sys/devices/system/cpu/cpu3/cpuidle/state*/name
POLL
C1-BDW
C1E-BDW
C3-BDW
C6-BDW
C7s-BDW
C8-BDW
C9-BDW
C10-BDW

---------------------------------------------------------------
Steve Best

Proposed Patch:
---------------

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/idle/intel_idle.c b/drivers/idle/intel_idle.c
index 10b86fa..5a5f1d8 100644
--- a/drivers/idle/intel_idle.c
+++ b/drivers/idle/intel_idle.c
@@ -395,6 +395,66 @@ static struct cpuidle_state hsw_cstates[] = {
  {
   .enter = NULL }
 };
+static struct cpuidle_state bdw_cstates[] = {
+ {
+  .name = "C1-BDW",
+  .desc = "MWAIT 0x00",
+  .flags = MWAIT2flg(0x00) | CPUIDLE_FLAG_TIME_VALID,
+  .exit_latency = 2,
+  .target_residency = 2,
+  .enter = &intel_idle },
+ {
+  .name = "C1E-BDW",
+  .desc = "MWAIT 0x01",
+  .flags = MWAIT2flg(0x01) | CPUIDLE_FLAG_TIME_VALID,
+  .exit_latency = 10,
+  .target_residency = 20,
+  .enter = &intel_idle },
+ {
+  .name = "C3-BDW",
+  .desc = "MWAIT 0x10",
+  .flags = MWAIT2flg(0x10) | CPUIDLE_FLAG_TIME_VALID | CPUIDLE_FLAG_TLB_FLUSHED,
+  .exit_latency = 40,
+  .target_residency = 100,
+  .enter = &intel_idle },
+ {
+  .name = "C6-BDW",
+  .desc = "MWAIT 0x20",
+  .flags = MWAIT2flg(0x20) | CPUIDLE_FLAG_TIME_VALID | CPUIDLE_FLAG_TLB_FLUSHED,
+  .exit_latency = 133,
+  .target_residency = 400,
+  .enter = &intel_idle },
+ {
+  .name = "C7s-BDW",
+  .desc = "MWAIT 0x32",
+  .flags = MWAIT2flg(0x32) | CPUIDLE_FLAG_TIME_VALID | CPUIDLE_FLAG_TLB_FLUSHED,
+  .exit_latency = 166,
+  .target_residency = 500,
+  .enter = &intel_idle },
+ {
+  .name = "C8-BDW",
+  .desc = "MWAIT 0x40",
+  .flags = MWAIT2flg(0x40) | CPUIDLE_FLAG_TIME_VALID | CPUIDLE_FLAG_TLB_FLUSHED,
+  .exit_latency = 300,
+  .target_residency = 900,
+  .enter = &intel_idle },
+ {
+  .name = "C9-BDW",
+  .desc = "MWAIT 0x50",
+  .flags = MWAIT2flg(0x50) | CPUIDLE_FLAG_TIME_VALID | CPUIDLE_FLAG_TLB_FLUSHED,
+  .exit_latency = 600,
+  .target_residency = 1800,
+  .enter = &intel_idle },
+ {
+  .name = "C10-BDW",
+  .desc = "MWAIT 0x60",
+  .flags = MWAIT2flg(0x60) | CPUIDLE_FLAG_TIME_VALID | CPUIDLE_FLAG_TLB_FLUSHED,
+  .exit_latency = 2600,
+  .target_residency = 7700,
+  .enter = &intel_idle },
+ {
+  .enter = NULL }
+};
 
 static struct cpuidle_state atom_cstates[] = {
  {
@@ -578,6 +638,11 @@ static const struct idle_cpu idle_cpu_hsw = {
  .disable_promotion_to_c1e = true,
 };
 
+static const struct idle_cpu idle_cpu_bdw = {
+ .state_table = bdw_cstates,
+ .disable_promotion_to_c1e = true,
+};
+
 static const struct idle_cpu idle_cpu_avn = {
  .state_table = avn_cstates,
  .disable_promotion_to_c1e = true,
@@ -604,7 +669,10 @@ static const struct x86_cpu_id intel_idle_ids[] = {
  ICPU(0x3f, idle_cpu_hsw),
  ICPU(0x45, idle_cpu_hsw),
  ICPU(0x46, idle_cpu_hsw),
- ICPU(0x4D, idle_cpu_avn),
+ ICPU(0x4d, idle_cpu_avn),
+ ICPU(0x3d, idle_cpu_bdw),
+ ICPU(0x4f, idle_cpu_bdw),
+ ICPU(0x56, idle_cpu_bdw),
  {}
 };
 MODULE_DEVICE_TABLE(x86cpu, intel_idle_ids);
-- 
1.7.1